### PR TITLE
Also fetch LAVA job definition for watchjobs

### PR DESCRIPTION
**Pull request by swvanbuuren**
Currently, the LAVA job definition is only fetched if the LAVA job is submitted directly through SQUAD (i.e. using a submitjob). This commit also implements this functionality for watchjobs.

### CODE REVIEW

**Comment by mwasilew** [comment]
LGTM

**Review by chaws — APPROVED** [review]
LGTM

File: squad/ci/backend/null.py [hunk 1]
```diff
@@ -92,18 +92,24 @@ def cancel(self, test_job):
...
         raise NotImplementedError
 
-    def check_job_definition(selt, definition):
+    def check_job_definition(self, definition):
         """
         Returns True if job definition checks out or an error message
         """
         raise NotImplementedError
 
+    def get_job_definition(self, test_job):
```
Comment (chaws) [inline squad/ci/backend/null.py:107]: Can you make this `job_id` instead of `test_job` as it only uses the job id?

File: squad/ci/models.py [hunk 2]
```diff
@@ -216,6 +216,9 @@ def get_implementation(self):
     def check_job_definition(self, definition):
         return self.get_implementation().check_job_definition(definition)
 
+    def get_job_definition(self, definition):
```
Comment (chaws) [inline squad/ci/models.py:219]: I think you meat `job_id` instead of `definition`?

File: squad/api/ci.py [hunk 3]
```diff
@@ -96,9 +96,14 @@ def watch_job(request, group_slug, project_slug, version, environment_slug):
     if testjob_id is None:
         return HttpResponseBadRequest("testjob_id is required")
 
+    definition = backend.get_job_definition(testjob_id)
+    if not definition:
```
Comment (chaws) [inline squad/api/ci.py:100]: There is another backend, [Tuxsuite](https://github.com/Linaro/squad/blob/master/squad/ci/backend/tuxsuite.py), and even though it **does not** have the concept of job definition, it still uses `/api/watchjob/` quite often. Returning a bad response wouldn't fit Tuxsuite use cases.

What do you think of this:
```diff
diff --git a/squad/api/ci.py b/squad/api/ci.py
index f59cd658..bb60adc0 100644
--- a/squad/api/ci.py
+++ b/squad/api/ci.py
@@ -96,14 +96,9 @@ def watch_job(request, group_slug, project_slug, version, environment_slug):
     if testjob_id is None:
         return HttpResponseBadRequest("testjob_id is required")
 
-    definition = backend.get_job_definition(testjob_id)
-    if not definition:
-        return HttpResponseBadRequest(f"No job definition found for job {testjob_id}: {definition}")
-
     # create TestJob object
     test_job = TestJob(
         backend=backend,
-        definition=definition,
         target=project,
         target_build=build,
         environment=environment_slug,
@@ -111,6 +106,16 @@ def watch_job(request, group_slug, project_slug, version, environment_slug):
         job_id=testjob_id
     )
 
+    # check if backend is able to fetch job definition
+    try:
+        definition = backend.get_job_definition(testjob_id)
+        if not definition:
+            return HttpResponseBadRequest(f"No job definition found for job {testjob_id}")
+        test_job.definition = definition
+    except NotImplementedError:
+        # backend does not work with job definitions
+        pass
+
     # sanitize job_url
     try:
         backend.get_implementation().job_url(test_job)
```

**Comment by chaws** [comment]
Tests are failing because the [testing backend](https://github.com/Linaro/squad/blob/master/squad/ci/backend/fake.py) is missing `get_job_definition()`. I came up with a quick fix in case you want to use it:
```diff
diff --git a/squad/ci/backend/fake.py b/squad/ci/backend/fake.py
index e49556bb..4b3100e2 100644
--- a/squad/ci/backend/fake.py
+++ b/squad/ci/backend/fake.py
@@ -66,5 +66,8 @@ class Backend(object):
     def check_job_definition(self, definition):
         return True
 
+    def get_job_definition(self, test_job):
+        return "sample job definition"
+
     def supports_callbacks(self):
         return False
diff --git a/test/api/test_ci.py b/test/api/test_ci.py
index 6ee36946..ba19fce3 100644
--- a/test/api/test_ci.py
+++ b/test/api/test_ci.py
@@ -238,6 +238,8 @@ class CiApiTest(TestCase):
             1,
             testjob_queryset.count()
         )
+        testjob = testjob_queryset.first()
+        self.assertTrue(len(testjob.definition) > 0)
         fetch.assert_called_with(testjob_queryset.first().id)
         logentry_queryset = LogEntry.objects.filter(
             user_id=self.project_privileged_user.pk,
```

File: squad/ci/backend/null.py (hunk 1 above)
Comment (swvanbuuren) [reply squad/ci/backend/null.py]: Fixed with [f28d3b7](https://github.com/Linaro/squad/pull/1123/commits/f28d3b75d0838a956d46611ba79c1dc640ff18ce)

File: squad/api/ci.py (hunk 3 above)
Comment (swvanbuuren) [reply squad/api/ci.py]: Yes, this makes sense. I have included the suggested change in commit [6eeec58](https://github.com/Linaro/squad/pull/1123/commits/6eeec58434b4aefa98c452645925716cc9197bd6)

**Comment by swvanbuuren** [comment]
> Tests are failing because the [testing backend](https://github.com/Linaro/squad/blob/master/squad/ci/backend/fake.py) is missing `get_job_definition()`. I came up with a quick fix in case you want to use it:
> 
> ```diff
> diff --git a/squad/ci/backend/fake.py b/squad/ci/backend/fake.py
> ...
> ```

Thank you for this. I have taken over your suggestion in commit [f81c12f](https://github.com/Linaro/squad/pull/1123/commits/f81c12f098c49059a088fb0219b954f2dec0e8cc).

Tests seem to have passed.

**Pull request merged**